lib/services/namespace_service.ts

```diff
@@ -0,0 +1,139 @@
+    this.deviceInterface!.setRawStorageValue(
+      NAMESPACES_STORAGE_KEY,
+      JSON.stringify(namespaces)
+    );
+  }
+
+  private async getDefaultNamespace() {
+    const namespaces = await this.getNamespaces();
+    return namespaces.find(namespace => namespace.isDefault === true);
+  }
+
+  private async pushNamespace(namespace: SNNamespace) {
+    const namespaces = await this.getNamespaces();
+    const namespaceIndex = namespaces.findIndex(n => n.identifier === namespace.identifier);
+    if (namespaceIndex === -1) {
+      namespaces.push(namespace);
+    } else {
+      namespaces[namespaceIndex] = namespace;
+    }
+    await this.setNamespaces(namespaces);
+  }
+
+  private async setCurrentNamespace(namespace: SNNamespace) {
+    this.namespace = namespace;
+    this.deviceInterface!.switchToNamespace(namespace);
+    await this.pushNamespace(namespace);
+  }
+
+  private async createNamespace(isDefault: boolean = false, label?: string) {
+    if (isDefault) {
+      const defaultNamespace = await this.getDefaultNamespace();
+      if (defaultNamespace) {
+        throw Error('Can not create default namespace: a default namespace already exists.');
+      }
+    }
+    const uuid = await Uuid.GenerateUuid();
+    const namespace: SNNamespace = {
+      identifier: uuid,
+      userUuid: undefined,
+      label: isDefault ? 'Default namespace' : label || uuid,
+      isDefault
+    };
+    return namespace;
+  }
+
+  /**
+   * Creates a new namespace if necessary, if not use an existing one.
+   */
+  public async initialize() {
+    const defaultNamespace = await this.getDefaultNamespace();
+    if (defaultNamespace) {
+      await this.setCurrentNamespace(defaultNamespace);
+      return;
+    }
+    const namespaces = await this.getNamespaces();
+    /** If no namespaces exist, then we should create a new one. */
+    if (namespaces.length === 0) {
+      const namespace = await this.createNamespace(true);
+      await this.setCurrentNamespace(namespace);
+    }
+    /**
+     * If one (1) namespace exist, then it means it is not the default one at this point.
+     * Let's set it as the default namespace and return it.
+     */
+    else if (namespaces.length === 1) {
+      const namespace = namespaces[0];
+      namespace.isDefault = true;
+      namespace.label = 'Default namespace';
+      await this.setCurrentNamespace(namespace);
+    }
+    /**
+     * In the future, if a user has multiple accounts signed into a client, each account
+     * will use its own keychain/storage namespace. This service will be in charge of displaying
+     * a UI for selecting which namespace to sign into (if no default namespace is set).
+     */
+    else if (namespaces.length > 1) {
+      throw Error('Multiple namespaces not supported yet ;)')
+    }
+  }
+
+  /**
+   * Creates a namespace with a fixed identifier.
+   * @param identifier The namespace identifier.
+   */
+  public createFixedNamespace(identifier: string) {
```

Comment:
Seems like this function should just be grouped in with `createNamespace`, where `createNamespace` would hae an optional last parameter called `identifier`. Also, it's not clear why this function calls `setCurrentNamespace` but createNamespace doesn't.